File: Tree/526-hasPathSum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:

        def f(root,s):
            if not root:
                return False
            
            s+=root.val
            if s==targetSum and not root.left and not root.right:
                return True
            return f(root.left,s) or f(root.right,s)
        return f(root,0)
    def hasPathSum2(self,root,target):
        def f(root,remainding):
            if not  root:
                return False
            remainding-=root.val
            
            if remainding==0 and not root.left and not root.right:
                return True
            return f(root.left,remainding) or f(root.right,remainding)
        return f(root,target)

File: Tree/526-hasPathSum.cpp
#include<iostream>
using std::cout;

class TreeNode{
    public:
        int val;
        TreeNode* left;
        TreeNode* right;
        TreeNode(int v,TreeNode* l=nullptr, TreeNode* r=nullptr):val{v},left{l},right{r}{};
};

bool f(const TreeNode* root, int v,int target)
{
    if(!root)
        return false;
    v+=root->val;
    if((v==target)&&(!root->left)&&(!root->right))
        return true;
    return f(root->left,v,target)||f(root->right,v,target);
}

bool hasPathSum(const TreeNode* root,int target)
{
    int iVal=0;
    return f(root,iVal,target);
}

int main()
{
    TreeNode* root=new TreeNode(7);
    TreeNode* one=new TreeNode(1);
    TreeNode* two=new TreeNode(2);
    TreeNode* three=new TreeNode(3);

    root->left=one;
    one->left=two;
    one->right=three;
    cout<<hasPathSum(root,11);
}
